### PR TITLE
 Add MarkupContent to HoverResponse

### LINKED ISSUE #141: Null on Hover result
Hello.

Currently i'm trying to create a lsp cli client for my student project. I'm dialoging with the clang language server `clangd`. I'm trying using your lsp types libraries but when i'm trying to get a `hover` result from the language server. I'm getting a null value but the language server give a result.
 
```json
{
  "id":4,
  "jsonrpc":"2.0",
  "result":{
      "contents":{
          "kind":"plaintext",
          "value":"Declared in global namespace\n\nint square(int x)"
      }
   }
}
```

I found the result given by `clangd` on my request and i found out the data `Hover` doesn't match the case where `contents` is `MarkupContent`.

```typescript
interface Hover {
	contents: MarkedString | MarkedString[] | MarkupContent;
	range?: Range;
}
```
 I create an edited `hover` data type to match this case.

```haskell
data Hover = Hover {
        _contents :: Either (List MarkedString) MarkupContent,
        _range    :: Maybe Range
    } deriving (Read, Show, Eq)

instance FromJSON Hover where
    parseJSON = withObject "object" $ \o -> do
        Hover <$> ( (Left  <$>  o .:  "contents") <|>
                            (Right <$> o .:  "contents") )
    
              <*>   o .:? "range"
```
I want to know if it's the right way to make it ?
Thank you.

### PR DESCRIPTION
**Pull request by alanz**
And deprecate the old MarkedString as per the LSP spec.

Closes #141

Addresses some of #134

### CODE REVIEW

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs [hunk 1]
```diff
@@ -1789,9 +1794,32 @@ instance FromJSON MarkedString where
...
+  parseJSON v@(A.String _) = HoverContentsMS <$> parseJSON v
+  parseJSON v@(A.Null)     = HoverContentsMS <$> parseJSON v
+  parseJSON v@(A.Array _)  = HoverContentsMS <$> parseJSON v
+  parseJSON v@(A.Object o) = do
+    mk <- o .:? "kind" :: Parser (Maybe MarkupKind)
+    case mk of
+      Nothing -> HoverContentsMS <$> parseJSON v
+      _       -> HoverContents   <$> parseJSON v
+  parseJSON _ = fail "HoverContents"
+
```
Comment (alanz) [inline haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: This `FromJSON` instance is horrible, I welcome any suggestions for improvement.

Comment (bubba) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: I used `Alternative` when dealing with the `CodeAction`/`Command` overload:

https://github.com/alanz/haskell-lsp/blob/cd94c9bef3b860b6187a3969b0e203a41c0ffef1/haskell-lsp-types/src/Language/Haskell/LSP/Types/CodeAction.hs#L285-L286

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/MarkupContent.hs [hunk 2]
```diff
@@ -120,3 +121,25 @@ data MarkupContent =
...
+-- | Create a 'MarkupContent' containing unquoted text
+unmarkedUpContent :: Text -> MarkupContent
+unmarkedUpContent str = MarkupContent MkPlainText str
+
+-- ---------------------------------------------------------------------
+
+instance Semigroup MarkupContent where
+  MarkupContent MkPlainText s1 <> MarkupContent MkPlainText s2 = MarkupContent MkPlainText (s1 <> s2)
+  MarkupContent MkMarkdown  s1 <> MarkupContent _           s2 = MarkupContent MkMarkdown  (s1 <> s2)
+  MarkupContent _           s1 <> MarkupContent MkMarkdown  s2 = MarkupContent MkMarkdown  (s1 <> s2)
```
Comment (bubba) [inline haskell-lsp-types/src/Language/Haskell/LSP/Types/MarkupContent.hs]: Nice touch 😃

File: haskell-lsp-types/haskell-lsp-types.cabal [hunk 3]
```diff
@@ -1,5 +1,5 @@
 name:                haskell-lsp-types
-version:             0.8.2.0
+version:             0.8.3.0
 synopsis:            Haskell library for the Microsoft Language Server Protocol, data types
```
Comment (bubba) [inline haskell-lsp-types/haskell-lsp-types.cabal]: Should this not be bumped to 0.9.0.0 since changing the `Hover` definition is a breaking change?

Comment (alanz) [reply haskell-lsp-types/haskell-lsp-types.cabal]: Ok

File: ChangeLog.md [hunk 4]
```diff
@@ -1,5 +1,9 @@
 # Revision history for haskell-lsp
 
+## 0.8.3.0
```
Comment (bubba) [inline ChangeLog.md]:
```suggestion
## 0.9.0.0
```

**Pull request merged**